### PR TITLE
CDRIVER-4255 Add asserts for functions with client and pool parameters

### PR DESCRIPTION
**Pull request by joshbsiegel**
- Looked up every instance of a function that has a `mongoc_client_t` or `mongoc_client_pool_t` parameter and added an assert to ensure that the respective function would not segfault when one of the parameters is `NULL`
- Decided to use `BSON_ASSERT_PARAM` so that the error message that will print out when the assert fails is specific to issue
- Used plain `ASSERT` in the test files
- In functions where one of the parameters I was looking for was marked as `BSON_UNUSED`, I did not add an assert as it seemed counterintuitive and there are some test cases where a client or pool is passed through to the function as NULL
- In functions where it was clear that the client or pool was not being dereferenced (such as one line functions that just call another functions), I did not add an assert

### CODE REVIEW

File: src/libmongoc/src/mongoc/mongoc-client-side-encryption.c [hunk 1]
```diff
@@ -88,6 +88,8 @@ void
 mongoc_auto_encryption_opts_set_keyvault_client (
    mongoc_auto_encryption_opts_t *opts, mongoc_client_t *client)
 {
+   BSON_ASSERT_PARAM (client);
```
Comment (kevinAlbs) [inline src/libmongoc/src/mongoc/mongoc-client-side-encryption.c:91]:
```suggestion
```

Suggest removing this assertion. A caller may be relying on passing `NULL` to clear the option.

File: src/libmongoc/src/mongoc/mongoc-client-side-encryption.c [hunk 2]
```diff
@@ -99,6 +101,8 @@ void
 mongoc_auto_encryption_opts_set_keyvault_client_pool (
    mongoc_auto_encryption_opts_t *opts, mongoc_client_pool_t *pool)
 {
+   BSON_ASSERT_PARAM (pool);
```
Comment (kevinAlbs) [inline src/libmongoc/src/mongoc/mongoc-client-side-encryption.c:104]:
```suggestion
```

Suggest removing this assertion. A caller may be relying on passing `NULL` to clear the option.

File: src/libmongoc/src/mongoc/mongoc-client-side-encryption.c [hunk 3]
```diff
@@ -266,6 +270,8 @@ void
 mongoc_client_encryption_opts_set_keyvault_client (
    mongoc_client_encryption_opts_t *opts, mongoc_client_t *keyvault_client)
 {
+   BSON_ASSERT_PARAM (keyvault_client);
+
```
Comment (kevinAlbs) [inline src/libmongoc/src/mongoc/mongoc-client-side-encryption.c:274]:
```suggestion
```

Suggest removing this assertion. A caller may be relying on passing `NULL` to clear the option.

File: src/libmongoc/src/mongoc/mongoc-opts.c [hunk 4]
```diff
@@ -23,6 +23,8 @@ _mongoc_insert_one_opts_parse (
 {
    bson_iter_t iter;
 
+   BSON_ASSERT_PARAM (client);
+
```
Comment (kevinAlbs) [inline src/libmongoc/src/mongoc/mongoc-opts.c:27]: mongoc-opts.c is generated by `build/generate-opts.py`. Suggest updating [mongoc-opts.c.template](https://github.com/mongodb/mongo-c-driver/blob/572fd74d70b2294deaff7e760d2c0175099cd331/build/opts_templates/mongoc-opts.c.template#L20) and regenerating mongoc-opts.c with `python ./build/generate-opts.py`.

File: src/libmongoc/tests/json-test-monitoring.c [hunk 5]
```diff
@@ -262,6 +262,7 @@ set_apm_callbacks (json_test_ctx_t *ctx, mongoc_client_t *client)
 {
    mongoc_apm_callbacks_t *callbacks;
 
+   ASSERT (client);
```
Comment (kevinAlbs) [inline src/libmongoc/tests/json-test-monitoring.c:265]: `client` is unused in this function. Suggest removing the `mongoc_client_t *client` argument.

File: src/libmongoc/src/mongoc/mongoc-opts.c (hunk 4 above)
Comment (joshbsiegel) [reply src/libmongoc/src/mongoc/mongoc-opts.c:27]: What should I be updating in the template? I searched for `ASSERT`s but didn't find any.

Comment (kevinAlbs) [reply src/libmongoc/src/mongoc/mongoc-opts.c:27]: Add `BSON_ASSERT_PARAM (client);` to the template that generates the function:

```
bool
_{{ struct_name }}_parse (
   mongoc_client_t *client,
   const bson_t *opts,
   {{ struct_type }} *{{ struct_name }},
   bson_error_t *error)
{
```

And regenerate mongoc-opts.c by running `python ./build/generate-opts.py`.

**Review by kevinAlbs — APPROVED** [review]
